Transition ActionPreference.Suspend enumeration value into a non-supported, reserved state, and remove restriction on using ActionPreference.Ignore in preference variables

KirkMunro opened the pull request:
# PR Summary

As part of the cleanup of workflow code, this PR does the following:
- Treats `ActionPreference.Suspend` enumeration value as not supported and reserved for future use.
- Update internal logic dependent on `ActionPreference.Suspend` such that using that value results in appropriate error messages.
- Remove unused resource strings.
- Updated and added tests related to `ActionPreference.Suspend` to ensure that it is generating the errors we expect when it is used.
- Remove unnecessary restriction around `ActionPreference.Ignore` (see [this discussion in issue 4348](https://github.com/PowerShell/PowerShell/issues/4348#issuecomment-490672315)). This change is made as part of this PR because it is in the exact same body of code, and it makes sense to include it here.

## PR Context

See issue #9570.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [X] Issue filed: [MicrosoftDocs/PowerShell-Docs #4624](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4624)
- **Testing - New and feature**
    - [X] N/A
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

KirkMunro committed:
remove deprecated ActionPreference.Suspend value

KirkMunro committed:
remove unused string

iSazonov commented on src/System.Management.Automation/engine/CommandBase.cs:
```diff
@@ -286,8 +286,6 @@ public enum ActionPreference
         /// <summary>Ignore the event completely (not even logging it to the target stream)</summary>
         Ignore,
-        /// <summary>Suspend the command for further diagnosis. Supported only for workflows.</summary>
-        Suspend,
```
I am not sure - that will break compiling with SDK.

KirkMunro replied:
I removed it based on [this comment](https://github.com/PowerShell/PowerShell/issues/9570#issuecomment-492842811). If necessary I can put it back, but that also would mean adding back logic to raise runtime errors if it is used. Better to remove it IMHO.

Also correct me if I'm wrong, but wouldn't breaking compiling with the SDK be a good thing in this case since workflow is out and the enumeration is no longer used, so code that uses it would need to be updated?

iSazonov replied:
> but that also would mean adding back logic to raise runtime errors if it is used.

No need to raise runtime errors. 
My thoughts is that the "Suspend" value is in public type and removing it will be a breaking change for scripts and codes. If we remove this we lost _binary_ backward compatibility too. 
There is 3 option: (1) Remove the value, (2) add Obsolete attribute, (3) add only comment that the value is "Obsolete".
Last option is my preference _today_ to avoid a breaking change with the public type. Similar used in CoreFX for some types so that we can continue to compile PowerShell without errors - it is not critical, related code is silently ignored. Also the CoreFX practice allows us to use new runtimes with previously compiled code.

KirkMunro replied:
I was thinking earlier this evening about the obsolete attribute. Seems with that I can decide if it generates compiler errors or warnings, right? If we generate warnings, that would preserve backwards compatibility while giving more visibility to the change, no?

iSazonov replied:
> If we generate warnings ...

It is not always so. If the user uses "warnings as errors" compiler option (we use it in the repo), then this does not work - the user will not be able to compile their project. That's why there is no use of this attribute in CoreFX.

KirkMunro replied:
The committee feedback was to keep the enum, so this has been addressed.

iSazonov reviewed (approved):
LGTM with one comment.

PaulHigin reviewed (approved):
LGTM

SteveL-MSFT commented:
@PowerShell/powershell-committee the decision here is what to do about removing a workflow specific value from the existing public enum.  One option is to mark it obsolete.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this, we discussed removal, using `[obsolete]` attribute, and leaving as-is.  We discussed the possibility of using this value in the future for job control if there is a way to suspend a PSJob.  For that reason, we agreed to leave this enum value.  We also recommend that we explicitly number enum values so in the future, if we do decide to remove a value and add another one, then existing code will break/behavior appropriately.

KirkMunro committed:
Merge branch 'master' into deprecate-actionpreference-suspend

KirkMunro committed:
put enum value back, updated plumbing and tests

KirkMunro commented:
To address the committee review feedback, I just made the following changes:

1. Put the enumerated value back.
1. Updated internal logic to generate appropriate errors about using an enumerated value that is not supported and reserved for future use.
1. Updated tests to ensure that it's behaving properly everywhere we need it to.

I'll also update the OP accordingly.

KirkMunro commented:
Note that regarding explicit numbering of the `ActionPreference` enumerated values, I have that handled in PR #9862 

Also, once either this PR or #8205 are merged in, I will update the other PR accordingly because there are some overlapping changes that I doubt will merge without manual intervention.

KirkMunro committed:
CodeFactor changes

PoshChan commented on src/System.Management.Automation/engine/ArgumentTypeConverterAttribute.cs:
```diff
@@ -147,15 +147,14 @@ internal object Transform(EngineIntrinsics engineIntrinsics, object inputData, b
                     // Note - this is duplicated in ExecutionContext.cs as parameter binding for script cmdlets can avoid this code path.
```
@KirkMunro, your last commit had 14 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Test-Connection.Ping.Default parameter set is 'Ping'
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 40
40:             $result = Test-Connection $targetName

```
Test-Connection.Ping.Count parameter
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 60
60:             $result1 = Test-Connection -Ping $targetName -Count 1

```
Test-Connection.Ping.Quiet works
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 68
68:             $result1 = Test-Connection $targetName -Count 1 -Quiet

```
Test-Connection.Ping.Force IPv4 with implicit PingOptions
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 89
89:             $result = Test-Connection $hostName -Count 1 -IPv4

```
Test-Connection.Ping.Force IPv4 with explicit PingOptions
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 100
100:             $result1 = Test-Connection $hostName -Count 1 -IPv4 -MaxHops 10 -DontFragment

```

PoshChan replied:
@KirkMunro, your last commit had 15 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Test-Connection.Ping.Default parameter set is 'Ping'
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 40
40:             $result = Test-Connection $targetName

```
Test-Connection.Ping.Count parameter
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 60
60:             $result1 = Test-Connection -Ping $targetName -Count 1

```
Test-Connection.Ping.Quiet works
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 68
68:             $result1 = Test-Connection $targetName -Count 1 -Quiet

```
Test-Connection.Ping.Force IPv4 with implicit PingOptions
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 89
89:             $result = Test-Connection $hostName -Count 1 -IPv4

```
Test-Connection.Ping.Force IPv4 with explicit PingOptions
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Test-Connection.Tests.ps1: line 100
100:             $result1 = Test-Connection $hostName -Count 1 -IPv4 -MaxHops 10 -DontFragment

```

PoshChan replied:
@KirkMunro, your last commit had 14 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Test-Connection.Ping.Default parameter set is 'Ping'
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 40
40:             $result = Test-Connection $targetName

```
Test-Connection.Ping.Count parameter
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 60
60:             $result1 = Test-Connection -Ping $targetName -Count 1

```
Test-Connection.Ping.Quiet works
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 68
68:             $result1 = Test-Connection $targetName -Count 1 -Quiet

```
Test-Connection.Ping.Force IPv4 with implicit PingOptions
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 89
89:             $result = Test-Connection $hostName -Count 1 -IPv4

```
Test-Connection.Ping.Force IPv4 with explicit PingOptions
```powershell
The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 100
100:             $result1 = Test-Connection $hostName -Count 1 -IPv4 -MaxHops 10 -DontFragment

```

KirkMunro committed:
change test to avoid failure (#8205 fixes this)

KirkMunro committed:
merge master

KirkMunro commented:
@iSazonov Could you please remove the Breaking Change label from this PR?

KirkMunro committed:
remove global qualifier added during testing

iSazonov commented:
@PaulHigin Could you please update your review (we have new commits)?

KirkMunro committed:
code review changes

KirkMunro commented:
FYI, I will not be making changes based on the Codacy review because one of the issues is a false positive, and the other two are incorrect. This PR is ready for review by other reviewers.

iSazonov commented:
@SteveL-MSFT PowerShell Committee request was addressed. Can we merge?

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1:
```diff
@@ -7,7 +7,7 @@ Describe "Test-Connection" -tags "CI" {
         $oldInformationPreference = $InformationPreference
         $oldProgressPreference = $ProgressPreference
-        $InformationPreference = "Ignore"
+        $InformationPreference = "SilentlyContinue"
```
Why does `Ignore` need to be updated to `SilentlyContinue` for this test?

KirkMunro replied:
The breaking change for `ActionPreference.Ignore` is very temporary, and a result of juggling a few related PRs while trying to keep their changes separate from one another (I've had such changes grouped in other PRs before and received pushback to separate them, so this is a result of that).

PR #10238 (which is waiting for review, hint, hint) removes the restriction around `ActionPreference.Ignore`. The removal of that restriction was approved in [this discussion in issue 4348](https://github.com/PowerShell/PowerShell/issues/4348#issuecomment-490672315).

Once we get both this PR and #10238 reviewed/merged, we'll be good.

daxian-dbw replied:
Thanks for pointing me to #4348. It makes sense to remove the check for `ActionPreference.Ignore`.
But I think you probably should incorporate the fix of `Remove unnecessary restriction around ActionPreference.Ignore` into this PR.

The PR #10238 is a feature change that seems not approved by the committee yet (I looked at  #6702 and your RFC), so it might take some time for it to get through.

KirkMunro replied:
I thought you might say that.

Related to PR #10238, see [this note from @SteveL-MSFT]([https://github.com/PowerShell/PowerShell/issues/6702#issuecomment-515147136]). That's not the committee, that's just Steve, but he did say "**we** would welcome a PR from the community", which certainly sounds like a welcome feature change to me.

Also I appreciate the suggestion that #10238 might take some time to get through. Ultimately I'm all for getting things done quickly, so if this PR must be held back because of the change being in that other PR, I'll move it over.

I should call out though that you collectively, as a team, request quite a bit of PR content juggling -- sometimes suggesting PRs should be combined, other times suggesting PRs should be split up, and others suggesting that content is moved from one PR to another. I try to make logical decisions with where PR changes go, ensuring that I only group like changes with other like changes, but the suggestions still come up. I'm really not pointing fingers here, and I'm not mad. I just feel I should call out that it's really not encouraging, and I feel it gets in the way of getting things done.

daxian-dbw replied:
I cannot speak for the overall guidelines/rules for contribution. I know @joeyaiello is actively working on that, and I'm sure it will be clearer and more transparent hopefully soon.
As a PR reviewer, I spot issues in the PR, and in this case, a breaking change. A breaking change will massively slow down the accepting of a PR because it will have to go through the committee again.
What I suggested here was a shortcut to get the PR merged faster, since once the minor fix is moved here, it will be an already-approved change given committee's review in https://github.com/PowerShell/PowerShell/issues/4348#issuecomment-490672315.

KirkMunro replied:
I agree the changes for `ActionPreference.Ignore` are better here. I wanted to avoid updating a few PRs, and was hoping to see both of them move forward, but since that's not likely I'll go for the route with the least amount of friction. I've brought the changes removing the restriction on `ActionPreference.Ignore` over here, making this non-breaking. I've also brought over the corresponding tests from the other PR, reverted the change to the `Test-Connection` test in this PR that was working around the breaking change, and will update the title and OP of this PR with the extra details in a moment.

You can remove the breaking change label now, @daxian-dbw.

KirkMunro replied:
As per the other comment, I brought over the change to get this merged in faster.

daxian-dbw commented on src/System.Management.Automation/engine/ExecutionContext.cs:
```diff
@@ -611,6 +602,11 @@ internal T GetEnumPreference<T>(VariablePath preferenceVariablePath, T defaultPr
+                    if (result is ActionPreference actionPreferenceValue)
+                    {
+                        CheckActionPreference(preferenceVariablePath, actionPreferenceValue, defaultPref);
+                    }
```
Technically, this is the right fix. This makes it consistent for `$InformationPreference = "Ignore"; Write-Information abc` to run in global scope (ps console directly) and in local scope.
However, this will also be a breaking change for scripts that accidentally use `$InformationPreference = "Ignore"` in local scopes ...

Today it behaves inconsistently:
```powershell
PS:48> $InformationPreference = "Ignore"; Write-Information abc ## error out !
Write-Information : The value Ignore is not supported for an ActionPreference variable. The provided value should be used only as a value for a preference parameter, and has been replaced by the default value. For more information, see the Help topic, "about_Preference_Variables."
At line:1 char:36
+ $InformationPreference = "Ignore"; Write-Information abc
+                                    ~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [Write-Information], NotSupportedException
+ FullyQualifiedErrorId : System.NotSupportedException,Microsoft.PowerShell.Commands.WriteInformationCommand

PS:49> & { $InformationPreference = "Ignore"; Write-Information abc }  ## no error ...
PS:50>
```

KirkMunro replied:
As per my previous comments, this is a very temporary breaking change. PR #10238 removed the unnecessary restriction around `ActionPreference.Ignore`, as per [this discussion in issue 4348](https://github.com/PowerShell/PowerShell/issues/4348#issuecomment-490672315).

Getting this PR and #10238 reviewed and merged in eliminates the breaking change that resulted in me changing a Pester test in this PR.

daxian-dbw replied:
I think that fix should be incorporate into this PR. I looks to me we just need to remove the check of `preference == ActionPreference.Ignore` in `GetEnumPreference`.

KirkMunro replied:
Done.

KirkMunro committed:
Merge branch 'master' into deprecate-actionpreference-suspend

KirkMunro committed:
remove ActionPreference.Ignore restriction
